src/libraries/Native/Unix/System.Native/pal_dl.c

```diff
@@ -0,0 +1,23 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#include "pal_dl.h"
+#include "dlfcn.h"
+#include <stdlib.h>
+
+#ifdef TARGET_ANDROID
+void* SystemNative_GetDefaultSearchOrderPseudoHandle(void)
+{
+    return (void*)RTLD_DEFAULT;
+}
+#else
+static void* g_defaultSearchOrderPseudoHandle = NULL;
+void* SystemNative_GetDefaultSearchOrderPseudoHandle(void)
+{
+    if (g_defaultSearchOrderPseudoHandle == NULL)
+    {
+        g_defaultSearchOrderPseudoHandle = dlopen(NULL, RTLD_LAZY);
```

Comment:
I assume dlopen properly handles concurrent usage and is idempotent?

Reply:
Mono implements this the same way so I believe it does for this case.

Reply:
I would be useful to double check this assumption.

Reply:
I've done some extended testing locally and `dlopen(NULL, RTLD_LAZY)` returns the same value every time since it's the pseudohandle.

Reply:
There is a slight cost per the [`dlopen(3)`](https://linux.die.net/man/3/dlopen) manpage:

> If the same library is loaded again with dlopen(), the same file handle is returned. The dl library maintains reference counts for library handles, so a dynamic library is not deallocated until dlclose() has been called on it as many times as dlopen() has succeeded on it. The _init() routine, if present, is only called once. But a subsequent call with RTLD_NOW may force symbol resolution for a library earlier loaded with RTLD_LAZY. 

so it is not completely free if we don't call `dlclose()` as many times (where possible; as in some cases we don't).

Reply:
I'm not sure that applies to the case where `NULL` is passed in. This isn't for a library but for the main program. The few implementations of `dlopen` I've looked at appear to special case `NULL` and return very early.

Reply:
Interesting that manpage does not specify the special leeway for `dlopen(NULL,..)` that we are free to skip `dlopen()`.
I only looked at glibc sources and it seems to at least increment/decrement reference counts (`l_direct_opencount`) regardless of `NULL`, as in; `dlclose()` on a handle returned by `dlopen(NULL)` is not a no-op.

Reply:
@am11 Agreed. I just confirmed your observation in glibc as well. Can't really argue with the fact that it does appear to increment and decrement, at least in the most common implementation. I did note the comment on the dummy for the executable, `_dl_main_map`, seems to indicate this is a very minimal instance. Perhaps the inc/dec doesn't impact this in the same way?

Reply:
I'm thinking `lt_executable` is the sentinel here.